routing/index_graph.hpp

```diff
@@ -28,33 +36,145 @@ class JointEdge final
 
 private:
-  // Target is vertex going to for outgoing edges, vertex going from for ingoing edges.
-  Joint::Id const m_target;
-  double const m_weight;
+  // Target is a vertex going to for outgoing edges, vertex going from for ingoing edges.
+  Joint::Id m_target;
+  double m_weight;
+};
+
+class JointEdgeGeom final
+{
+public:
+  JointEdgeGeom() = default;
+
+  JointEdgeGeom(Joint::Id target, vector<RoadPoint> const & path) : m_target(target), m_path(path)
+  {
+  }
+
+  Joint::Id GetTarget() const { return m_target; }
+
+  vector<RoadPoint> const & GetPath() const { return m_path; }
+
+private:
+  // Target is a vertex going to for outgoing edges, vertex going from for ingoing edges.
+  Joint::Id m_target = Joint::kInvalidId;
+  vector<RoadPoint> m_path;
+};
+
+class DirectedEdge final
+{
+public:
+  DirectedEdge() = default;
+  DirectedEdge(Joint::Id from, Joint::Id to, uint32_t featureId)
+    : m_from(from), m_to(to), m_featureId(featureId)
+  {
+  }
+
+  bool operator<(DirectedEdge const & rhs) const
+  {
+    if (m_from != rhs.m_from)
+      return m_from < rhs.m_from;
+
+    if (m_to != rhs.m_to)
+      return m_to < rhs.m_to;
+
+    return m_featureId < rhs.m_featureId;
+  }
+
+  bool operator==(DirectedEdge const & rhs) const
+  {
+    return m_from == rhs.m_from && m_to == rhs.m_to && m_featureId == rhs.m_featureId;
+  }
+
+  Joint::Id GetFrom() const { return m_from; }
+  Joint::Id GetTo() const { return m_to; }
+  uint32_t GetFeatureId() const { return m_featureId; }
+
+private:
+  Joint::Id const m_from = Joint::kInvalidId;
+  Joint::Id const m_to = Joint::kInvalidId;
+  // Note. It's important to store feature id because two |m_from| and |m_to| may be
+  // connected with several features.
+  uint32_t const m_featureId = 0;
+};
+
+inline bool IsAdjacent(DirectedEdge const & ingoing, DirectedEdge const & outgoing)
+{
+  return ingoing.GetTo() == outgoing.GetFrom();
+}
+
+class RestrictionInfo final
+{
+public:  
+  RestrictionInfo() = default;
+
+  RestrictionInfo(DirectedEdge const & ingoing, DirectedEdge const & outgoing)
+    : m_from(ingoing.GetFrom()), m_center(ingoing.GetTo()), m_to(outgoing.GetTo()),
+      m_fromFeatureId(ingoing.GetFeatureId()), m_toFeatureId(outgoing.GetFeatureId())
+  {
+    CHECK(IsAdjacent(ingoing, outgoing), ());
+  }
+
+  bool operator<(RestrictionInfo const & rhs) const
+  {
+    if (m_from != rhs.m_from)
+      return m_from < rhs.m_from;
+
+    if (m_center != rhs.m_center)
+      return m_center < rhs.m_center;
+
+    if (m_to != rhs.m_to)
+      return m_to < rhs.m_to;
+
+    if (m_fromFeatureId != rhs.m_fromFeatureId)
+      return m_fromFeatureId < rhs.m_fromFeatureId;
+
+    return m_toFeatureId < rhs.m_toFeatureId;
+  }
+
+  bool operator==(RestrictionInfo const & rhs) const { return !(*this < rhs || rhs < *this); }
+
+  bool operator!=(RestrictionInfo const & rhs) const { return !(*this == rhs); }
+
+  pair<DirectedEdge, DirectedEdge> ToEdges()
```

Comment:
A bit confusing ToEdges - it is 'convert it to edges', but also it can be treated - 'To' as a second edge. 
May be rename it as a just 'edges()'?